File: 719-find-k-th-smallest-pair-distance/find-k-th-smallest-pair-distance.py
class Solution:
    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        nums.sort()
        l, r = 0, max(nums)
        while l < r:
            m = l + ((r - l) // 2)
            count = 0
            left = 0
            for right in range(1, len(nums)):
                while nums[right] - nums[left] > m and left <= right:
                    left += 1
                count += right - left
            if count < k:
                l = m + 1
            else:
                r = m
        return r

File: 719-find-k-th-smallest-pair-distance/find-k-th-smallest-pair-distance.cpp
class Solution {
public:
    int smallestDistancePair(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int l = 0, r = nums.back();
        while (l < r){
            int m = l + ((r - l) / 2);
            int count = 0, left = 0;
            for (int right = 1; right < nums.size(); ++right){
                while (nums[right] - nums[left] > m && left <= right){
                    left++;
                }
                count += right - left;
            }
            if (count >= k){
                r = m;
            }else {
                l = m + 1;
            }
        }
        return r;
    }
};